src/aiida_quantumespresso/workflows/xspectra/core.py

```diff
@@ -0,0 +1,910 @@
+        shell_inputs['nodes'] = {'upf': upf}
+        shell_inputs['arguments'] = orm.List(list=['{upf}'])
+        shell_inputs['metadata'] = {'call_link_label': 'upf2plotcore'}
+
+        shelljob_node = self.submit(ShellJob, **shell_inputs)
+        self.report(f'Launching ShellJob for upf2plotcore.sh<{shelljob_node.pk}>')
+
+        return ToContext(upf2plotcore_node=shelljob_node)
+
+    def inspect_upf2plotcore(self):
+        """Check that the output from the upf2plotcore step has yielded a meaningful result.
+
+        This will simply check that the core wavefunction data returned contains at least
+        one core state and return an error if this is not the case.
+        """
+
+        shelljob_node = self.ctx.upf2plotcore_node
+        core_wfc_data = shelljob_node.outputs.stdout
+        header_line = shelljob_node.outputs.stdout.get_content()[:40]
+        num_core_states = int(header_line.split(' ')[5])
+        if num_core_states == 0:
+            return self.exit_codes.ERROR_NO_GIPAW_INFO_FOUND
+        self.ctx.core_wfc_data = core_wfc_data
+
+    def run_all_xspectra_prod(self):
+        """Run an `XspectraBaseWorkChain` for each 3-vector given for epsilon."""
+
+        eps_vectors = self.inputs.eps_vectors.get_list()
+        parent_folder = self.ctx.scf_workchain.outputs.remote_folder
+        core_wfc_data = self.ctx.core_wfc_data
+
+        xspectra_prod_calcs = {}
+        xspectra_calc_labels = []
+        calc_number = 0
+        for vector in eps_vectors:
+            xspectra_inputs = AttributeDict(self.exposed_inputs(XspectraBaseWorkChain, 'xs_prod'))
+            xspectra_parameters = xspectra_inputs.xspectra.parameters.get_dict()
+
+            parent_folder = self.ctx.scf_workchain.outputs.remote_folder
+            xspectra_inputs.xspectra.parent_folder = parent_folder
+            xspectra_inputs.xspectra.core_wfc_data = core_wfc_data
+
+            label = f'xas_{calc_number}'
+            xspectra_inputs.metadata.call_link_label = f'{label}_prod'
+            xspectra_inputs.metadata.label = f'{label}_prod'
+            xspectra_calc_labels.append(label)
+
+            for index in [0, 1, 2]:
+                xspectra_parameters['INPUT_XSPECTRA'][f'xepsilon({index + 1})'] = vector[index]
+            xspectra_inputs.xspectra.parameters = orm.Dict(dict=xspectra_parameters)
+
+            if self.ctx.dry_run:
+                return xspectra_inputs
+
+            future_xspectra = self.submit(XspectraBaseWorkChain, **xspectra_inputs)
+            self.report(
+                f'launching XspectraWorkChain<{future_xspectra.pk}> for epsilon vector {vector}'
+                ' (Lanczos production)'
+            )
+            xspectra_prod_calcs[f'{label}_prod'] = future_xspectra
+            calc_number += 1
+
+        if 'xspectra_calc_labels' not in self.ctx.keys():
+            self.ctx.xspectra_calc_labels = xspectra_calc_labels
+        return ToContext(**xspectra_prod_calcs)
```

Comment:
Instead of manually creating labels and adding them to the context as this, if you just need a list, you can use the following in the for-loop as well
```python
from aiida.engine import append_

for index, vector in enumerate(eps_vectors):
    ...
    future_xspectra = self.submit(XspectraBaseWorkChain, **xspectra_inputs)
    self.to_context(xspectra_calculations=append_(future_xspectra)
```
This will automatically create a list in the `ctx` under the key `xspectra_calculations` to which the xspectra nodes are appended. You can then use this in `inspect_all_xspectra_prod` as follows:
```python
for xspectra_calculation in self.ctx.xspectra_calculations`
```
Think that makes for a lot more readable code